components/backup-stream/src/subscription_manager.rs

```diff
@@ -485,10 +446,106 @@ where
                     }
                     callback(ResolvedRegions::new(rts, cps));
                 }
+                ObserveOp::HighMemUsageWarning {
+                    inconsistent_region_id,
+                } => {
+                    self.on_high_memory_usage(inconsistent_region_id).await;
+                }
+            }
+        }
+    }
+
+    async fn on_observe_result(
+        &mut self,
+        region: Region,
+        handle: ObserveHandle,
+        err: Option<Box<Error>>,
+    ) {
+        let err = match err {
+            None => {
+                self.failure_count.remove(&region.id);
+                let sub = self.subs.get_subscription_of(region.id);
+                if let Some(mut sub) = sub {
+                    if sub.value().handle.id == handle.id {
+                        sub.value_mut().resolver.phase_one_done();
+                    }
+                }
+                return;
             }
+            Some(err) => {
+                if !should_retry(&err) {
+                    self.failure_count.remove(&region.id);
+                    self.subs
+                        .deregister_region_if(&region, |sub, _| sub.handle.id == handle.id);
+                    return;
+                }
+                err
+            }
+        };
+
+        let region_id = region.id;
+        *self.failure_count.entry(region_id).or_insert(0) += 1;
+        match self.retry_observe(region.clone(), handle).await {
+            Ok(has_resent_req) => {
+                if !has_resent_req {
+                    self.failure_count.remove(&region_id);
+                }
+            }
+            Err(e) => {
+                self.issue_fatal_of(
+                    &region,
+                    e.context(format_args!(
+                        "retry encountered error, origin error is {}",
+                        err
+                    )),
+                );
+                self.failure_count.remove(&region_id);
+            }
+        }
+    }
+
+    async fn on_high_memory_usage(&mut self, inconsistent_region_id: u64) {
+        let mut lame_region = Region::new();
+        lame_region.set_id(inconsistent_region_id);
+        let mut act_region = None;
+        self.subs.deregister_region_if(&lame_region, |act, _| {
+            act_region = Some(act.meta.clone());
+            true
+        });
+        let delay = OOM_BACKOFF_BASE
+            + Duration::from_secs(rand::thread_rng().gen_range(0..OOM_BACKOFF_JITTER_SECS));
+        info!("log backup triggering high memory usage."; 
+            "region" => %inconsistent_region_id, 
+            "mem_usage" => %self.memory_manager.used_ratio(), 
+            "mem_max" => %self.memory_manager.capacity());
+        if let Some(region) = act_region {
+            self.schedule_start_observe(delay, region, None);
         }
     }
 
+    fn schedule_start_observe(
```

Comment:
Make it `async fn` then we don't need `tokio::spawn(async move {}) at line 540.

Reply:
Even this become an async fn we still cannot await in this function because it may be executed on the event loop.